File: moe/natj/natj/src/test/java/org/moe/xosrt/binding/test/directCallTest/testClasses/DirectCallClassWBlocks.java
/*
Copyright 2014-2016 Intel Corporation

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package org.moe.xosrt.binding.test.directCallTest.testClasses;


import org.moe.natj.general.NatJ;
import org.moe.natj.general.NativeObject;
import org.moe.natj.general.Pointer;
import org.moe.natj.objc.ann.ObjCBlock;
import org.moe.natj.objc.ann.ObjCClassName;
import org.moe.natj.objc.ann.Selector;
import apple.NSObject;
import org.moe.xosrt.binding.test.directCallTest.testClasses.Blocks.BlockMultipleTwoFloats;
import org.moe.xosrt.binding.test.directCallTest.testClasses.Blocks.BlockObject;
import org.moe.xosrt.binding.test.directCallTest.testClasses.Blocks.CharBlockRetObject;
import org.moe.xosrt.binding.test.directCallTest.testClasses.Blocks.FloatBlock;
import org.moe.xosrt.binding.test.directCallTest.testClasses.Blocks.FloatBlockRetFloat;
import org.moe.xosrt.binding.test.directCallTest.testClasses.Blocks.FloatBlockRetObject;
import org.moe.xosrt.binding.test.directCallTest.testClasses.Blocks.ObjectBlockRetObject;
import org.moe.xosrt.binding.test.directCallTest.testClasses.Blocks.ObjectOnlyBlock;
import org.moe.xosrt.binding.test.directCallTest.testClasses.Blocks.SimpleBlock;

@ObjCClassName("DirectCallClassWBlocks")
public class DirectCallClassWBlocks extends NativeObject {
	
	static
	{
		NatJ.register();
	}

	protected DirectCallClassWBlocks(Pointer peer) {
		super(peer);
	}

	@Selector("alloc")
	public static native DirectCallClassWBlocks alloc(); 
	
	@Selector("init")
	public native DirectCallClassWBlocks init();
	
	@Selector("DirectCallWBlock:")
	public native void DirectCallWBlock(@ObjCBlock(name = "BlockCallback") SimpleBlock block);
	
	
	@Selector("DirectCallDerniBlock")
	public native void DirectCallDerniBlock();

	@Selector("DirectCallWBlockFloat:")
	public native void DirectCallWBlockFloat(@ObjCBlock(name = "BlockCallBack") FloatBlock block);
	
	@Selector("DirectCallWBlockObject:withObject:andSObj:andTObj:")
	public native void DirectCallWBlockObject(@ObjCBlock(name = "ObjectMethod") BlockObject block, NativeObject object,
			NativeObject sObject, NativeObject tObject);
	
	@Selector("DirectCallWObjectOnlyBlock:withObject:andSObj:andTObj:")
	public native void DirectCallWObjectOnlyBlock(@ObjCBlock(name = "BlockCallBack") ObjectOnlyBlock block, NativeObject object,
			NativeObject sObject, NativeObject tObject);
	
	@Selector("DirectCallWObjectBlockRetObject:withObject:")
	public native NSObject DirectCallWObjectBlockRetObject(@ObjCBlock(name = "BlockCallBack") ObjectBlockRetObject block, NSObject object);
	
	@Selector("DirectCallFloatBlockRetObject:fFloat:sFloat:withObject:")
	public native NSObject DirectCallFloatBlockRetObject(@ObjCBlock(name = "BlockCallBack") FloatBlockRetObject block, float a, float b, NativeObject object);
	
	@Selector("DirectCallCharBlockRetObject:charArg:byteArg:withObject:")
	public native NSObject DirectCallCharBlockRetObject(@ObjCBlock(name = "BlockCallBack") CharBlockRetObject block, char a, byte b, NSObject object);
		
	@Selector("DirectCallFloatBlockRetFloat:fFloat:byteArg:withObject:sFloat:")
	public native float DirectCallFloatBlockRetFloat(@ObjCBlock(name = "BlockCallBack") FloatBlockRetFloat block, float a, byte b, NativeObject object, float c);
	
	@Selector("setCompletionBlock:")
	public native void setCompletionBlock(@ObjCBlock(name = "BlockCallback") SimpleBlock block);
	
	@Selector("DirectCallWithBlock:FFloat:SFloat:")
	public native float DirectCallWithBlockMultipleTwoFloats(@ObjCBlock(name = "BlockCallBack") BlockMultipleTwoFloats ptr, float a, float b);
	
	
	
}

File: natj/natj/src/test/native/DirectCallTest/DirectCallClassWBlocks.m
/*
Copyright 2014-2016 Intel Corporation

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/


#import <Foundation/Foundation.h>
#import "DirectCallClassWBlocks.h"
#import "test_common.h"


@implementation DirectCallClassWBlocks : NSObject

@synthesize completionBlock;


-(void) DirectCallWBlock:(void (^)()) argBlock
{
    argBlock();
}

-(void) DirectCallWBlockFloat:(void (^)(float a, float b)) argBlock
{
    TESTLog(@"Trying to call block in DirectCallWBlockFloat");
    argBlock(10, 20);
}

-(void) DirectCallWBlockObject:(void (^)(int a, double s,  NSObject *object, int b, double z, NSObject *secObject, NSObject *thirdObject)) argBlock
                    withObject: (NSObject *)fobject
                       andSObj:(NSObject *)sobject
                       andTObj:(NSObject *)tobject
{
    
    //TESTLog(@"Trying to call block in DirectCallWBlockFloat");
    argBlock(15 ,100,fobject, 27, 350,  sobject, tobject);
}

-(void) DirectCallWObjectOnlyBlock:(void (^)(NSObject *object, NSObject *secObject, NSObject *thirdObject)) argBlock
                        withObject:(NSObject *)fobject
                           andSObj:(NSObject *)sobject
                           andTObj:(NSObject *)tobject
{
    argBlock(fobject,sobject, tobject);
}

-(NSObject *) DirectCallWObjectBlockRetObject:(NSObject* (^)(NSObject *object)) argBlock
                                   withObject:(NSObject *)fobject
{
   NSObject* newObject = argBlock(fobject);
    
    TESTLog(@"NEW Object: %@", newObject);
    
    return newObject;
}

-(NSObject *) DirectCallFloatBlockRetObject:(NSObject* (^)(float a, float b, NSObject *object)) argBlock fFloat:(float) aArg sFloat:(float) bArg
                                   withObject:(NSObject *)fobject
{
    
    TESTLog(@"DirectCallFloatBlockRetObject:%f:%f:%@", aArg, bArg, fobject);
    NSObject* newObject = argBlock(aArg, bArg,fobject);
    TESTLog(@"NEW Object: %@", newObject);
    
    return newObject;
}

- (NSObject *) DirectCallCharBlockRetObject:(NSObject* (^)(char a, Byte b, NSObject *object)) argBlock charArg:(char) aArg byteArg:(Byte) bArg
withObject:(NSObject *)fobject
{
    return argBlock(aArg,bArg,fobject);
}

- (float) DirectCallFloatBlockRetFloat:(float (^)(float a, Byte b, NSObject *object, float c)) argBlock fFloat:(float) aArg byteArg:(Byte) bArg
                                 withObject:(NSObject *)fobject sFloat:(float) cArg
{
    return argBlock(aArg,bArg,fobject, cArg);
}

-(void) DirectCallDerniBlock
{
    completionBlock();
}

-(float) DirectCallWithBlock:(float(^)(float a, float b))block FFloat:(float) ffloat SFloat:(float) sfloat
{
    return block(ffloat, sfloat);
}

@end
